### DISCUSSION TITLE
어떤게 더 좋은 건가요?( if or func)

### DISCUSSION THREAD

**Discussion by Lcat12345** [MyeongWoonJang/cpp-study#149] [Q&A] [answered]
조건문이 길어지는게 나은가요,
함수 호출이 많아지는게 나은가요.

조건문이 길어진다는 것은

```
if ( 괄호안에 들어가는 조건들이 길어지거나 )
  if( )
    if( )
     {
중첩 if문이 되는 경우를 말하고
}
```

함수 호출이 많아 진다는 것은
```
if ( func() ){}

func() { return func2(); } ... 
```
이런식으로 함수호출이 계속 되는 경우를 말합니다.

**Comment by Woon-2** [comment, marked as answer]
`function`이 많아지는 것이 좋습니다.

`maintainability` 측면에서 `function call`이 많아지는 것이 훨씬 우월하고요.   
`performance` 측면에서 `function call`이 많아지는 걸로 `loss`를 보지 않을 확률이 높습니다.

## maintainability 측면

우리가 `high level programming language`를 이용함으로써 얻는 이득들 중에는   
의미 단위로 `coding`이 가능하다는 점도 있습니다.

어떠한 일련의 동작들에 의미를 부여해서 하나의 `function`을 만들면,   
구체적인 동작을 외울 필요 없이 그 `function`의 의미만 기억하면 되죠.

책을 외울 때 목차를 먼저 떠올리고 항목별 내용을 떠올리는 것,   
비슷한 종류의 파일들을 묶어 한 폴더 안에 정리하는 것,   
열거된 내용을 외울 때 앞글자만 따서 외우는 것 등등   
의미 단위를 만드는 것은 많은 양의 정보를 효과적으로 다루는 제일의 방법입니다.

서로 다른 `subfunction`들의 조합으로 나타낼 수 있는 `function`을    
단순히 `variable`들과 `operation`들의 나열로 표현하면,   
`function`의 내부 동작들에 어떠한 의미도 부여하지 않겠다는 말입니다.   

날것의 `variable`과 `operation`들은 그 개수가 엄청나게 많을 것입니다.
그리고 각각의 `variable`에 대한 `access`가 잘못되지는 않았는지,   
잘못된 `operation`을 적용하고 있지는 않는지 등의 고민을    
엄청나게 긴 `function body`를 읽는 내내 해야 합니다.   

서로 분할할 수 있는 작업들이 결합되어있는 경우에도 그래요.   
`A` 블록과 `B` 블록의 작업들이 분할된 작업이어도,   
`A` 블록에서 이용하는 `variable`을 `B` 블록에서 `write`하고 있지는 않는지 검사해야 하는 일이 실제로 발생합니다.    
한 `function`에서 이유 모를 `bug`가 발생하면,   
그 `function body` 전체에 대해 `side effect`가 발생하지는 않았는지,   
`logic`이 잘못되지는 않았는지 검사해야 할테니까요.   

분명히 작업을 하다보면 `function`의 특정한 부분만 재활용하고싶을 때가 있을텐데,   
그 부분에서 쓰는 `variable`들의 `declaration`은 `function`의 저 윗부분에 있고   
`variable`들의 `declaration`과 재활용하고 싶은 문장들 사이에 일종의 `dependency`가 있는 문장이 있을 수도 있습니다.

어떻게 어떻게 문제가 없도록 잘 골라내서 `copy&paste`했는데,   
문제가 없어 "보였고", 실제로 문제가 있었다면요?   
"도대체 어디가 잘못된 거야~" 하고 골치아파질 겁니다.   
또한, 이미 그 부분이 재활용이 여러 번 되었고, 뒤늦게야 문제가 발견된 것이라면
여태까지 `copy&paste`했던 모든 부분의 `code`를 점검 및 수정해야 합니다.   

문제는 없었는데, `requirement`가 변경되었다면요?    
예를 들어 `ASCII` 문자열만 쓰던 `code`였는데, 갑자기 `UTF-16` 문자열에 대한 호환성이 필요해진 겁니다.   
`code`상에 쓰인 모든 `char*`를 찾아 일일이 `char16_t*`를 이용하는 `code`들로 변환할 건가요?   
`char*`가 쓰인 곳 중에서는 해당 `requirement`와 관계 없는 곳도 있을 텐데요...

`high level programming language programmer`라면 최대한 명확하고 분할된 의미소를 생성하세요.   
`function`과 `class`를 통해서요.      
사실 `code duplication`을 제거하라는 역사적이고 꽤나 강압적인 조언의 실천 방법이 이것입니다.

## performance 측면

다른 `programming language`라면 `performance loss`가 `C++`에 비해 클 수 있습니다.   

허나 `C++`는 `inline`이라는 강력한 기능 덕분에   
`function call`처럼 보이는 문장이 실제로 `function call`이 아닐 수 있어요.

그래서 `function call`로 인한 `overhead`를 완전히 줄일 수 있습니다.   
마치 `function call`을 이용하지 않은 것처럼요.

단지 `inline`에만 집중해보세요.   
일부러 어려운 `example code`를 뽑아왔습니다.   

```cpp
template<length_t L, typename T, qualifier Q>
GLM_FUNC_QUALIFIER vec<L, T, Q> fastNormalize(vec<L, T, Q> const& x)
{
	return x * fastInverseSqrt(dot(x, x));
}
```

`fastNormalize`는 `subfunction`으로 `fastInverseSqrt`와 `dot`를 이용합니다.

간단해보이는 `dot`의 `inline`만 살펴봅시다.

```cpp
template<length_t L, typename T, qualifier Q>
GLM_FUNC_QUALIFIER GLM_CONSTEXPR T dot(vec<L, T, Q> const& x, vec<L, T, Q> const& y)
{
	GLM_STATIC_ASSERT(std::numeric_limits<T>::is_iec559, "'dot' accepts only floating-point inputs");
	return detail::compute_dot<vec<L, T, Q>, T, detail::is_aligned<Q>::value>::call(x, y);
}
```

`dot`는 `subfunction`으로 `detail::compute_dot::call`을 이용하네요.
`function`의 이름이 `call`인 걸 보니 `detail::compute_dot`는 `function class` 즉 `functor`일 것 같습니다.

```cpp
namespace detail
{
	template<typename V, typename T, bool Aligned>
	struct compute_dot{};

        /**
         * 4차원 벡터에 대한 템플릿 특수화만 불린다고 가정하고
         * 1~3차원 벡터들에 대한 템플릿 특수화는 생략하겠습니다.
         * 지금 중요한 건 템플릿 특수화가 아니니깐요.
         */
	template<typename T, qualifier Q, bool Aligned>
	struct compute_dot<vec<1, T, Q>, T, Aligned>
	{ ... };

	template<typename T, qualifier Q, bool Aligned>
	struct compute_dot<vec<2, T, Q>, T, Aligned>
	{ ... };

	template<typename T, qualifier Q, bool Aligned>
	struct compute_dot<vec<3, T, Q>, T, Aligned>
	{ ... };
        /*****************************************************************/

        // 이 특수화가 불린다고 가정해봅시다.
	template<typename T, qualifier Q, bool Aligned>
	struct compute_dot<vec<4, T, Q>, T, Aligned>
	{
		GLM_FUNC_QUALIFIER GLM_CONSTEXPR static T call(vec<4, T, Q> const& a, vec<4, T, Q> const& b)
		{
			vec<4, T, Q> tmp(a * b);
			return (tmp.x + tmp.y) + (tmp.z + tmp.w);
		}
	};
}
```

모든 `template function`(`function template` 말고요.)은 `implicitly inline`입니다.   
거꾸로 올라가서 `function`들이 어떻게 `inline`되는지 봅시다.

먼저, `detail::compute_dot::call`의 `inline`입니다.
```cpp
template<length_t L, typename T, qualifier Q>
GLM_FUNC_QUALIFIER GLM_CONSTEXPR T dot(vec<L, T, Q> const& x, vec<L, T, Q> const& y)
{
	GLM_STATIC_ASSERT(std::numeric_limits<T>::is_iec559, "'dot' accepts only floating-point inputs");

	/**
         * return detail::compute_dot<vec<L, T, Q>, T, detail::is_aligned<Q>::value>::call(x, y);
         * ^--- 이 코드가 아래와 같이 바뀐 것처럼 동작합니다.
         */
	vec<4, T, Q> tmp(x * y);
	return (tmp.x + tmp.y) + (tmp.z + tmp.w);
}
```

그리고 `dot`의 `inline`입니다.  

```cpp
template<length_t L, typename T, qualifier Q>
GLM_FUNC_QUALIFIER vec<L, T, Q> fastNormalize(vec<L, T, Q> const& x)
{
	/**
         * return x * fastInverseSqrt(dot(x, x));
         * ^--- 이 코드가 아래와 같이 바뀐 것처럼 동작합니다.
         */
        GLM_STATIC_ASSERT(std::numeric_limits<T>::is_iec559, "'dot' accepts only floating-point inputs");
        return x * fastInverseSqrt(detail::compute_dot<vec<L, T, Q>, T, detail::is_aligned<Q>::value>::call(x, x));
}
```

아직 `inline`이 한 번 더 남았죠?   
`detail::compute_dot::call`가 `inline`되어야 하잖아요?

```cpp
template<length_t L, typename T, qualifier Q>
GLM_FUNC_QUALIFIER vec<L, T, Q> fastNormalize(vec<L, T, Q> const& x)
{
        GLM_STATIC_ASSERT(std::numeric_limits<T>::is_iec559, "'dot' accepts only floating-point inputs");
	/**
         * return x * fastInverseSqrt(detail::compute_dot<vec<L, T, Q>, T, detail::is_aligned<Q>::value>::call(x, x));
         * ^--- 이 코드가 아래와 같이 바뀐 것처럼 동작합니다.
         */
	vec<4, T, Q> tmp(x * x);
	return x * fastInverseSqrt((tmp.x + tmp.y) + (tmp.z + tmp.w));
}
```

사실, 두 벡터의 곱셈 연산자(`operator*`)도 따로 `function`으로 구현되어 있을 것이기 때문에
`inline` 과정이 몇 개 더 숨겨져 있었습니다. (제일 밑단이죠.)   

글이 길어지니 한번에 살펴볼까요?

```cpp
namespace detail
{
	template<typename T, qualifier Q, bool Aligned>
	struct compute_vec4_mul
	{
		GLM_FUNC_QUALIFIER GLM_CONSTEXPR static vec<4, T, Q> call(vec<4, T, Q> const& a, vec<4, T, Q> const& b)
		{
			return vec<4, T, Q>(a.x * b.x, a.y * b.y, a.z * b.z, a.w * b.w);
		}
	};
}

template<typename T, qualifier Q>
template<typename U>
GLM_FUNC_QUALIFIER GLM_CONSTEXPR vec<4, T, Q>& vec<4, T, Q>::operator=(vec<4, U, Q> const& v)
{
	this->x = static_cast<T>(v.x);
	this->y = static_cast<T>(v.y);
	this->z = static_cast<T>(v.z);
	this->w = static_cast<T>(v.w);
	return *this;
}

template<typename T, qualifier Q>
template<typename U>
GLM_FUNC_QUALIFIER GLM_CONSTEXPR vec<4, T, Q> & vec<4, T, Q>::operator*=(vec<4, U, Q> const& v)
{
        // 바로 위의 vec<4,T,Q>::operator=과 그 위의 detail::compute_vec4_mul::call을 inline합니다.
	return (*this = detail::compute_vec4_mul<T, Q, detail::is_aligned<Q>::value>::call(*this, vec<4, T, Q>(v)));
}

template<typename T, qualifier Q>
GLM_FUNC_QUALIFIER GLM_CONSTEXPR vec<4, T, Q> operator*(vec<4, T, Q> const& v1, vec<4, T, Q> const& v2)
{
	return vec<4, T, Q>(v1) *= v2;   // 바로 위의 vec<4, T, Q>::operator*=을 inline합니다.
}
```

자 다시 최초의 `fastNormalize`로 돌아가서 `dot`가 어떻게 `inline`되었을지 살펴봅시다.   

```cpp
template<length_t L, typename T, qualifier Q>
GLM_FUNC_QUALIFIER vec<L, T, Q> fastNormalize(vec<L, T, Q> const& x)
{
        GLM_STATIC_ASSERT(std::numeric_limits<T>::is_iec559, "'dot' accepts only floating-point inputs");

	vec<4, T, Q> tmp(x);
        vec<4, T, Q> mul_result(tmp.x * x.x, tmp.y * x.y, tmp.z * x.z, tmp.w * x.w);
        tmp.x = mul_result.x;
        tmp.y = mul_result.y;
        tmp.z = mul_result.z;
        tmp.w = mul_result.w;

	return x * fastInverseSqrt((tmp.x + tmp.y) + (tmp.z + tmp.w));
}
```

수많은 `function call`이 사라지고 위와 같은 `code`만 남게 되었습니다.   
분리할 수 있는 의미소들이 날것 그대로 녹아있는 코드죠?   
`operation`들이 겉보기에는 뭔가 비효율적인 게 눈에 보입니다만, 그쪽은 제쳐두죠.    

이런 것이 `inline`입니다.
"우리 눈에 보이는 `code`"와 "실제 `compile`된 `code`"를 달리 취함으로써    
`maintainability`와 `performance` 두 마리 토끼를 잡는 데에 도움을 줄 수 있는 도구이지요.

모든 `function`이 `virtual function`인 몇몇 `programming language`들에서는 상상도 할 수 없습니다.   
`compile time`에 어떤 `function`이 `call`될지 결정할 수 없기 때문에,   
(=`run time`에 어떤 `function`이 `call`될지 결정하기 때문에)    
`function call` 부분을 `function body`로 치환할 수가 없거든요.

`function call` 한 개 정도의 `overhead`는 미미하지만,   
보통 작업하다 보면 `function call`이 연쇄적으로 이루어지는 경우가 많습니다.   
그리고 그것이 `performance bottleneck(성능 병목)`이 되는 경우가 존재하죠.

하지만, `C++`은 `inline`의 수혜를 얻을 수 있는 `programming language`이므로,   
**`function` 이용의 장점인 의미 부여는 가져가면서 `function` 이용의 단점인 `call overhead`는 안 가져가는**   
**이기적인 `programming`이 가능합니다.**

의미는 우리 눈에 보이는 `code`에만 있으면 되거든요.
